cli/flox-rust-sdk/src/models/search.rs

```diff
@@ -290,42 +299,123 @@ impl TryFrom<&[u8]> for SearchResults {
+                let sender = sender.clone();
+                let deserializer = serde_json::Deserializer::from_reader(stdout);
+                for maybe_record in deserializer.into_iter() {
+                    let record = maybe_record.map_err(SearchError::Deserialize)?;
+                    if let Err(err) = sender.send(PkgDbOutput::Stdout(record)) {
+                        debug!("failed to send stdout line: {err}");
+                    }
+                }
+                trace!("stdout reader thread is done");
+                Ok(())
+            });
+
+            // Read items from the channel in the order in which they were received
+            let mut count = None;
+            let mut results = Vec::new();
+            while let Ok(output) = receiver.recv() {
+                match output {
+                    PkgDbOutput::Stderr(line) => {
+                        debug!(target: "pkgdb", "[pkgdb] {}", line);
+                    },
+                    PkgDbOutput::Stdout(Record::Error(err)) => {
+                        debug!("error from pkgdb: {}", err);
+                        let kill = pkgdb_process.kill().map_err(SearchError::PkgDbCall);
+                        // This destructuring is necessary for type conversion
+                        if let Err(err) = kill {
+                            return Ok(Err(err));
+                        }
+                        if !stderr_thread.is_finished() {
+                            trace!("waiting for stderr thread to finish");
+                            let _ = stderr_thread.join();
+                        }
+                        if !stdout_thread.is_finished() {
+                            trace!("waiting for stdout thread to finish");
+                            let _ = stdout_thread.join();
+                        }
+                        return Ok(Err(SearchError::PkgDb(err)));
+                    },
+                    PkgDbOutput::Stdout(Record::ResultCount { result_count }) => {
+                        debug!("result count = {}", result_count);
+                        count = Some(result_count);
+                    },
+                    PkgDbOutput::Stdout(Record::SearchResult(result)) => {
+                        debug!("search result = {:?}", result);
+                        results.push(result);
+                    },
+                }
+            }
+
+            trace!("joining reader threads");
+            let stderr_thread_outcome = stderr_thread.join();
+            let stdout_thread_outcome = stdout_thread.join();
+            trace!("done joining reader threads");
+            if stderr_thread_outcome.is_ok() && stdout_thread_outcome.is_ok() {
+                Ok(Ok((results, count)))
+            } else {
+                Ok(Err(SearchError::SomethingElse(
+                    "background threads didn't exit successfully".into(),
+                )))
+            }
```

Comment:
Actually, if you already handle the `panic` cases here, you dont even need to return a `std::thread::Result` (see how we never construct an `Err(_)` variant.

further suggestions (nb)

Also `SomethingElse` can then become specific i.e. `ThreadPanicked`.

We can then try downcasting the error into a `Box<dyn Error>` ?> `Box<dyn Display>` ?> `Box<dyn Debug>`  to revive a bit moe information that is otherwise lost.

Reply:
I mean it was functional from the beginning, but yeah I'd like to move on :)